src/Discord.Net.WebSocket/DiscordSocketClient.cs

```diff
@@ -627,29 +631,51 @@ private async Task ProcessUserDownloadsAsync(IEnumerable<SocketGuild> guilds)
         {
             var cachedGuilds = guilds.ToImmutableArray();
 
-            const short batchSize = 1;
-            ulong[] batchIds = new ulong[Math.Min(batchSize, cachedGuilds.Length)];
-            Task[] batchTasks = new Task[batchIds.Length];
-            int batchCount = (cachedGuilds.Length + (batchSize - 1)) / batchSize;
+            foreach (var guild in cachedGuilds)
+            {
+                await ApiClient.SendRequestMembersAsync(guild.Id).ConfigureAwait(false);
+                await guild.DownloaderPromise.ConfigureAwait(false);
+            }
+        }
 
-            for (int i = 0, k = 0; i < batchCount; i++)
+        /// <inheritdoc />
+        public override async Task DownloadUsersAsync(IGuild guild, IEnumerable<ulong> userIds, CancellationToken cancelToken = default)
+        {
+            if (ConnectionState == ConnectionState.Connected)
             {
-                bool isLast = i == batchCount - 1;
-                int count = isLast ? (cachedGuilds.Length - (batchCount - 1) * batchSize) : batchSize;
+                EnsureGatewayIntent(GatewayIntents.GuildMembers);
 
-                for (int j = 0; j < count; j++, k++)
+                var socketGuild = GetGuild(guild.Id);
+                if (socketGuild != null)
                 {
-                    var guild = cachedGuilds[k];
-                    batchIds[j] = guild.Id;
-                    batchTasks[j] = guild.DownloaderPromise;
+                    foreach (var chunk in userIds.Chunk(DiscordConfig.MaxRequestedUserIdsPerRequestGuildMembersChunk))
+                    {
+                        await ProcessUserDownloadsAsync(socketGuild, chunk, cancelToken).ConfigureAwait(false);
+                    }
                 }
+            }
+            else
+            {
+                throw new InvalidOperationException("Client not connected");
+            }
+        }
 
-                await ApiClient.SendRequestMembersAsync(batchIds).ConfigureAwait(false);
 
-                if (isLast && batchCount > 1)
-                    await Task.WhenAll(batchTasks.Take(count)).ConfigureAwait(false);
-                else
-                    await Task.WhenAll(batchTasks).ConfigureAwait(false);
+        private async Task ProcessUserDownloadsAsync(SocketGuild guild, IEnumerable<ulong> userIds, CancellationToken cancelToken = default)
+        {
+            var nonce = Interlocked.Increment(ref _guildMembersRequestCounter).ToString();
+            var tcs = new TaskCompletionSource<bool>();
+            using var registration = cancelToken.Register(() => tcs.TrySetCanceled());
+            _guildMembersRequestTasks.TryAdd(nonce, tcs);
+            try
+            {
+                await ApiClient.SendRequestMembersAsync(guild.Id, userIds, nonce).ConfigureAwait(false);
+                await tcs.Task.ConfigureAwait(false);
```

Comment:
the cancellation token should be integrated with the await of the tcs

Reply:
that's done a few lines above with the cancellation token registration, unless I'm missing something